controllers/parser.go

```diff
@@ -0,0 +1,357 @@
+
+    Independently, each virtual cluster in the virtual host is checked, in order. If there is a match, the virtual cluster is used and no further virtual cluster checks are made.
+
+From the istio issue tracker:
+
+The virtual hosts order does not influence the domain matching order
+
+It is the domain matters
+
+Domain search order:
+1. Exact domain names: www.foo.com.
+2. Suffix domain wildcards: *.foo.com or *-bar.foo.com.
+3. Prefix domain wildcards: foo.* or foo-*.
+4. Special wildcard * matching any domain.
+*/
+
+// UpdateConfigFromAPIOpts updates Envoy configuration from OpenAPI spec and x-kusk options
+func UpdateConfigFromAPIOpts(envoyConfiguration *config.EnvoyConfiguration, opts *options.Options, spec *openapi3.T) error {
+	for _, vhost := range opts.Hosts {
+		vh := types.NewVirtualHost(string(vhost))
+		envoyConfiguration.AddVirtualHost(vh)
+	}
+
+	// Iterate on all paths and build routes
+	// The overriding works in the following way:
+	// 1. For each path we get SubOptions from the opts map and merge in top level SubOpts
+	// 2. For each method we get SubOptions for that method from the opts map and merge in path SubOpts
+	for path, pathItem := range spec.Paths {
+		// x-kusk options per operation (http method)
+		for method, operation := range pathItem.Operations() {
+
+			finalOpts := opts.OperationFinalSubOptions[method+path]
+			if finalOpts.Disabled != nil && *finalOpts.Disabled {
+				continue
+			}
+
+			routePath := path
+			if finalOpts.Path != nil {
+				routePath = generateRoutePath(finalOpts.Path.Prefix, path)
+			}
+
+			corsPolicy, err := generateCORSPolicy(finalOpts.CORS)
+			if err != nil {
+				return err
+			}
+
+			rt := &route.Route{
+				Name: generateRouteName(routePath, method),
+				Match: generateRouteMatch(
+					routePath,
+					method,
+					extractParams(operation.Parameters),
+					corsPolicy,
+				),
+			}
+
+			// This block creates redirect route
+			// We either create the redirect or the route with proxy to upstream
+			// Redirect takes a precedence.
+			if finalOpts.Redirect != nil {
+				routeRedirect, err := generateRedirect(finalOpts.Redirect)
+				if err != nil {
+					return err
+				}
+
+				rt.Action = routeRedirect
+			} else {
+				upstreamHostname, upstreamPort := getUpstreamHost(finalOpts.Upstream)
+				clusterName := generateClusterName(upstreamHostname, upstreamPort)
+				if !envoyConfiguration.ClusterExist(clusterName) {
+					envoyConfiguration.AddCluster(clusterName, upstreamHostname, upstreamPort)
+				}
+
+				routeRoute, err := generateRoute(
+					clusterName,
+					corsPolicy,
+					&generateRouteOpts{
+						RewriteRegex: &opts.Path.Rewrite,
+						QoS:          opts.QoS,
+					},
+				)
+				if err != nil {
+					return err
+				}
+
+				rt.Action = routeRoute
+			}
+
+			for _, vh := range envoyConfiguration.GetVirtualHosts() {
+				if err := envoyConfiguration.AddRouteToVHost(vh.Name, rt); err != nil {
+					return fmt.Errorf("failure adding redirect route: %w", err)
```

Comment:
"redirect route" while it is just a generic route now (i.e. could be a redirect or a route)